Adds tracking amount of time spent playing departments, and refactors job unlock to use it,

Shadowlight213 opened the pull request:
The I have no idea what I'm doing anymore edition! Now with hopefully an expandable tracking system.
Take 2 since oranges wanted me to make a new pr
Previous pr #26088

@MrStonedOne 
:cl:
experiment: The amount of time spent playing, and departments played are now tracked per player.
experiment: This tracking can be used as a requirement of playtime to unlock jobs.
/:cl:


Basically every 10 minutes all active players get time added for the role they are playing.
Roles are tracked for: living, crew, each department, silicon, ghost, and antag.

Also adds support for basing the time before being able to play jobs to be based off time played jobs.
Time amounts are in minutes and were just ported from the original pr.

TODO:
  - [x] Make exp earning work with the new storage method
- [x] Make the time be saved to a list and only do the sql updates at round end.
 - [x] Add button to let admins toggle exempt status
- [x] Table mekhi
- [x] Update the database changelog
- [x] Local testing
- [ ] Final testing
 
~~Look into doing the low priority lockout thing~~ Probably will be in a second pr.
 ~~Port the paradise expansion stuff~~ Just wanting to get this over with

Shadowlight213 committed:
Ports Paradise Job exp system

Shadowlight213 committed:
changed to use json_decode and json_encode as recommended by oranges

Shadowlight213 committed:
updates changelog

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

Shadowlight213 committed:
Reee cyberboss

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

Shadowlight213 committed:
Changes to use a seperate table

Shadowlight213 committed:
Updates database changelog
Changes to use mediumint

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

Shadowlight213 committed:
Back to wip we go.
Changes sql stuff to how mso suggested
Attempts the changes requested in the other pr.

Repukan commented:
Are old players going to be grandfathered in?

MrStonedOne commented:
No, what i'd likely do is enable tracking, but make it not enforce, then after some time, enable enforcement.

Repukan commented:
Hmm seems fair, but for roles like HoP, whose departments under their belt prepare nothing for the head role, how would that go?

Would there be a mixed requirement for captains?

Shadowlight213 commented:
@MrStonedOne enforcement is a config
Also yes. there's an sql command to grandfather everyone in while keeping tracking data

MrStonedOne commented:
shadow, what i'd likely do is exempt everybody that had played in the last few months, then in a few months, remove the exemptions.

MrStonedOne commented:
And i'm going to come out and say it now, this has **massive fucking value**.

When I was a new player, you know kept me from saying "fuck the rules I do what I want" the most?

I wanted to play as AI, and that took a 30 day investment, and I had assumed at the time that only days I played count, so i wouldn't be able to pre-prime accounts.

It doesn't have to be much, but if you give accounts value, you cut down on rule disregard.

Repukan commented:
Oh yeah definitely, I'd think anyone would hardly object to having head roles locked behind experience. But again, for captain and HoP roles, which roles count towards their unlock?

Repukan commented:
And like would you need to put certain hours into sec officer, then unlock and play warden, and then play it to unlock HoS?

Would you only need experience as a normal engineer or as both Atmos and normal engineering to unlock CE?

Shadowlight213 commented:
@Repukan only the department.
So you could just play all the time as engi and unlock ce
same with sec officer to unlock hos.

Hop requires hours playing cargo roles
Captain doesn't have any departmental requirement. I could make them require playing heads of staff though

Repukan commented:
Ja that seems nice.

Make assistant players gain negative time for shit and giggles.

Durkel commented:
I kind of find a certain level of new player incompetence endearing. We're also implying that you're somehow becoming more competent the more time you put into the game, but It's pretty self evident that's not nearly the case with how often the fucking SM shard blows or what medical looks like daily.

Fox-McCloud commented:
Generalized playtime tracking is a good thing.

Departmental is NOT a good thing as enumerated in the conversation in the original PR: https://github.com/ParadiseSS13/Paradise/pull/7020 for a large number of reasons.

Implementing playtime tracking is a step in the right direction; tracking individual playtime in departments to make decisions based on that? Bad.

MrStonedOne commented:
Sadly, it has to be done.

Could just make it restrict them to low priority after some amount of general play time.

Fox-McCloud commented:
You can implement tracking on a global level, then lock jobs based on global playtime; but per department has a number of problems associated with it, the least of which is accuracy. 

I'm not cautioning against global playtime tracking (based on when you're crew, that is), just basing it on a per-department scenario.

Fox-McCloud commented:
@MrStonedOne 

You can implement tracking on a global level, then lock jobs based on global playtime; but per department has a number of problems associated with it, the least of which is accuracy. 

I'm not cautioning against global playtime tracking (based on when you're crew, that is), just basing it on a per-department scenario.

Basically, what I'm saying is, this: https://github.com/ParadiseSS13/Paradise/pull/5224 is a good PR.

This one, however, has a large number of problems with it: https://github.com/ParadiseSS13/Paradise/pull/7020 and I would not recommend it, at all.

Shadowlight213 commented:
Mso specifically requested departmentalized tracking when I first suggested porting this.

MrStonedOne commented:
Look at what i said again

You can make it so that after they reach a certain amount of play time, they unlock low priority on jobs they would not have enough of normally.

Even add it as something you get with byond memebership by default.

RemieRichards commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,186 @@
+// Procs
+
+
+/datum/job/proc/available_in_playtime(client/C)
```
bad name.
```required_playtime_remaining(client/C)```

```available_in_playtime(client/C)``` reads like it returns a bool, it doesn't, it returns 0-SOME AMOUNT, which actually if you were to use it like a bool (as you've done in preferences) would be the opposite of how it appears!

```if(available_in_playtime(client))``` will only succeed if it's NOT available!

tl;dr bad name.

kevinz000 commented:
?!

CPTANT commented:
Ok I am going to be seriously annoyed when I suddenly can't play AI or HoS because I haven't been playing that much lately.

FantasticFwoosh commented:
Do you have your department scout badge in clowning before becoming HOP?

Yeah we do have some roles like service that might become redundant or filled with low-skill pubbies in this regard if they aren't moved onto sec or another role. (mysterious new head?)

xxalpha commented:
So rarely plays veterans need to grind game time to play the jobs they like?

KorPhaeron commented:
This is already a difficult game to get into without making people grind for unlocks
 
Especially coming from other servers.

Our game already requires an admin presence to run, so I dont know why automated, collective, pre-emptive punishment is required rather than just job bans.

Also like @Durkel said incompetence is part of the flavour of the game and setting. We don't need perfect control and tracking to create ideally functioning stations

KorPhaeron commented:
1 hour to unlock med doctor

10 hours to CMO

10 hours to HoS

10 hours to RD

10 hours to CE

10 hours to HoP

51 hours at minimum to get all the jobs

MrStonedOne commented:
Because it adds value to your account's standing at /tg/

and like I said, we can easily do something like unlock all at low once one is unlocked, or allow byond memebers to get low status out of the gate.

KorPhaeron commented:
I don't think putting up hurdles to rule abiding players to try and bribe ban evaders to play nice is worth it

MrStonedOne commented:
I do

KorPhaeron commented:
This will do far more damage to the average players experience than the rounds ruined by ban evaders.

I guess this is going to boil down to a generic freedom vs security argument though

MrStonedOne commented:
Like I said, there are many things we can do to alleviate your concerns.

Abusing low job priority is a really good way to do that.

also, having to unlock headroles is such a minor thing, nobody is really going to care they can't be the head of the department, they can still play the department otherwise.

but lets have a poll?

KorPhaeron commented:
Can run a poll if you want, but I'm sure it would come out heavily in favour of this (though I think the majority would be motivated by trying to prevent incompetent heads rather than ban evaders)

MrStonedOne commented:
`Question: If you had to have play time in a department before you could play that department's head, how much play time would you find acceptable before you would see it as off putting?`

MrStonedOne commented:
http://www.ss13.eu/tgdb/tg/ingamepolls.php#p167

PKPenguin321 commented:
Good for playtime tracking but locking jobs is questionable

xxalpha commented:
That poll does not include "everyone starts from 0 hours counted".

MrStonedOne commented:
Thats not whats beings measured, xxalpha. Carefully re-read the question. I worded it very intentionally because i'm just trying to prove a point to kor about a specific case, new players finding the system off putting

MrStonedOne commented:
Anywho, @Shadowlight213, maybe change it to 3 hours based on the poll, we can raise it later on.

As we raise it we can gauge new user reaction to it, and maybe start adding other limits, like for ai, scientist, security, etc.

KorPhaeron commented:
Also wait, if I kill someone I stop their XP timer?

I understand you don't want people afking as ghost for their points but that is a pretty substantial OOC punishment for getting killed

KorPhaeron commented on code/modules/jobs/job_types/engineering.dm:
```diff
@@ -14,6 +14,9 @@ Chief Engineer
 	selection_color = "#ffeeaa"
 	req_admin_notify = 1
 	minimal_player_age = 7
+	exp_requirements = 600
```
These should be config based and default to 0 for downstreams who mirror our PRs

Shadowlight213 replied:
There already is a config for overriding the time needed for head of staff jobs.
Also, tracking time, using the restriction for head of staff jobs, using it for non head of staff jobs, and requiring department specific playtime are all separate configs.

MrStonedOne commented:
You aren't playing the department and learning it if you aren't alive now are you?

Repukan commented:
>mfw being unrobust now carries ooc penalties.

KorPhaeron commented:
Yeah but now I am griefing people on an OOC level when I kill them. Murdersprees are annoying enough without them adding time to your pseudo job bans

MrStonedOne commented:
a newbie is more likely to get killed by fucking up their internals then a antag.

Shadowlight213 commented:
Ok. This is the times for each job as it stands before I do any changes. These values were a direct port from the original pr.
The departmental requirements are for any job in that department. so for CE, it could be any combination of 10 hours playing atmos tech and/or engineer.

Captain: 10 hours crew
CMO: 10 hours as medical job
Medical doctor: 1 hour as crew
Genetecist: 1 hour crew
Chemist: 3 hours crew
Viro: 3 hours crew
HOS: 10 hours sec department
Sec officer: 5 hours crew
Warden: 5 hours crew
Detective: 5 hours crew
RD: 10 hours science department
Scientist: 1 hour crew
Robotocist: 1 hour crew
CE: 10 hours engineering
Engineer: 3 hours crew
Atmos Tech: 3 hours crew
HOP: 10 hours cargo
AI: 10 hours crew
Borg: 5 hours crew

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

Shadowlight213 committed:
work on stuff.

Shadowlight213 committed:
work on stuff.

Shadowlight213 committed:
why did I try to do a smart

PKPenguin321 commented:
Heavy dislike for job locking, takes way too long for most jobs

Shadowlight213 committed:
Renamed proc for time remaining.
Fixed updating time amounts.
Added delay to fix firing at roundstart

Shadowlight213 committed:
exp is now stored in a global list, and only updated to the db at round end.

MrStonedOne commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+	for(var/client/L in GLOB.clients)
+		if(L.inactivity >= 6000)
+			continue
+		addtimer(L.update_exp_list(mins, ann),10)
```
this isn't how add timer works

Cyberboss replied:
You want `addtimer(CALLBACK(L, /client/proc/update_exp_list, mins, ann), 10)`

MrStonedOne commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/category in GLOB.exp_jobsmap)
```
See, this isn't how you so should do this.

Don't add the exp to the different categories, save the exp by job/type, then have checking the exp add up the relevant jobs/types.

As it is, if you make a new exp requirement, it wouldn't track. if you just track all jobs, as one, you don't have to worry about that down the line. (eg, say you add a new job, call it the troll, and make it require clown and/or mime exp, well it would have no data under this system, but if you just tracked all jobs and roles, regardless, new roles and requirements down the line would have data) (Also @nfreader would love the extra data even if no jobs require that job)

Shadowlight213 replied:
exp_jobsmap has nothing to do with job unlock requirements. It basically groups tracked info into categories.
Unless you're saying you want me to track every single job seperately?

MrStonedOne replied:
I do

MrStonedOne replied:
The system should track the jobs, all jobs, and the requirements should just merge the relevant jobs when it checks for them.

This is more expansion friendly.

Shadowlight213 replied:
okay. that's um.
I mean I can do that, but that's an awful lot of data to be pulling from the db for every player on connect.

MrStonedOne replied:
lets say 30 jobs, each job is 40 characters, and there are 120 players reconnecting after a reboot.

That's still only ~150kb.

MrStonedOne replied:
And with primary key indexing on the two columns, ckey and job, it will be a very fast lookup.

Lets say that all 50,000 players in the player table have a record, lets say each player has 30 jobs recorded.

50000 players + 30 job = 1,500,000 rows

Mysql uses a sorted binary tree for primary key indexes, to do the query to find every job row for a ckey, since the ckey is the first column in the pimary key index, it will do a binary search for the first matching ckey row in the index then it grab all the rows that follow until it gets to the end of that ckey.

A binary search is O(log2(n)), so at 1,500,000 rows, it will find the first one after 21 lookups.

So thats 2,520 lookups **total**, for all 120 people reconnecting.

In comparison, every time a client connects, a row is inserted in the connection_log database, and this triggers a sorted insert of like 5 columns into indexes, each having to do 23 look ups, so thats 115 lookups per connection, or 13,000 lookups total for 120 players reconnecting at reboot.

Shadowlight213 replied:
Whoops! I forgot to update the list proc to the new system when I readded it. My bad.

MrStonedOne commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+		for(var/jtype in GLOB.exp_to_update[keyname])
+			var jobname = jtype
+			var time = GLOB.exp_to_update[keyname][jtype]
+			var/datum/DBQuery/update_query = SSdbcore.NewQuery("INSERT INTO [format_table_name("role_time")] (`ckey`, `job`, `minutes`) VALUES ('[sanitizeSQL(keyname)]', '[sanitizeSQL(jobname)]', '[sanitizeSQL(time)]') ON DUPLICATE KEY UPDATE minutes = '[sanitizeSQL(time)]'")
```
if they play on server 1, die after 10 minutes, then play on server 2 and get an hour of play time, all of that will be lost. if server 1 updates its round after server 2.
You can not assume your stored data is still good.

The best way would be to do this when you update everybody's exp, as += queries

Also, you can merge this into one big query easy, rather then do it in a loop. i'll make a generic function for doing so and get back to you.

when you do ON DUPLICATE KEY UPDATE, you should also always use `column = VALUES(column)` so that it pulls from the values and you don't have to send the same data twice, it also helps when you do multiple row inserts since it just pulls the data that would have been inserted.

Cyberboss commented on SQL/tgstation_schema.sql:
```diff
@@ -254,6 +254,21 @@ CREATE TABLE `messages` (
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8 */;
+
+CREATE TABLE `SS13_role_time` 
```
This is the prefixed schema

Cyberboss commented on code/controllers/configuration.dm:
```diff
@@ -104,6 +104,13 @@
+	var/use_exp_restrictions_heads_hours = 0
+	var/use_exp_restrictions_heads_department = 0
+	var/use_exp_restrictions_other = 0
+	var/use_exp_restrictions_admin_bypass = 0
```
FALSE

Cyberboss commented on code/controllers/configuration.dm:
```diff
@@ -329,6 +336,18 @@
+				if("use_exp_restrictions_other")
+					use_exp_restrictions_other = 1
+				if("use_exp_restrictions_admin_bypass")
+					use_exp_restrictions_admin_bypass = 1
```
TRUE

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+GLOBAL_LIST_INIT(exp_to_update, list())
```
GLOBAL_LIST_EMPTY(exp_to_update)

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+	var/body = "<html><head><title>Playtime for [C.key]</title></head><BODY><BR>Playtime:"
+	body += C.get_exp_report()
+	body += "</BODY></HTML>"
+	usr << browse(body, "window=playerplaytime[C.ckey];size=550x615")
```
Make body a list and use body.Join() in the browse call

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+	for(var/client/C in GLOB.clients)
+		msg += "<LI> - [key_name_admin(C)]: <A href='?_src_=holder;getplaytimewindow=\ref[C.mob]'>" + C.get_exp_living() + "</a></LI>"
+	msg += "</UL></BODY></HTML>"
+	src << browse(msg, "window=Player_playtime_check")
```
See my next comment

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+	else
+		return "0h"
+
+/proc/update_exp(var/mins, var/ann = 0)
```
This entire proc should be inside server_maint/fire

Shadowlight213 replied:
See reply on the blackbox ss comment

Cyberboss commented on code/controllers/subsystem/blackbox.dm:
```diff
@@ -29,6 +35,11 @@ SUBSYSTEM_DEF(blackbox)
 
+	if(config.use_exp_tracking)
+		if((triggertime < 0) || (world.time > (triggertime +3000)))	//server maint fires once at roundstart then once every 10 minutes. a 5 min check skips the first fire. The <0 is midnight rollover check
+			update_exp(10,0)
```
This should be done by the server_maint SS. It's wait was reduced to 6 a while back unlike this comment implies

Shadowlight213 replied:
Nope. Wont work. Server maint having a wait of 6 causes it to fire far too often for this. You changed it from 10 minutes making it useless for what I needed.

Cyberboss commented on code/controllers/subsystem/blackbox.dm:
```diff
@@ -97,6 +108,7 @@ SUBSYSTEM_DEF(blackbox)
 
 	var/datum/DBQuery/query_feedback_save = SSdbcore.NewQuery("INSERT DELAYED IGNORE INTO [format_table_name("feedback")] VALUES " + sqlrowlist)
 	query_feedback_save.Execute()
+	update_exp_db()
```
Also move to server_maint

Shadowlight213 replied:
You mean shutdown, right?

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+		play_records = prefs.exp
+		if(!play_records.len)
+			return "[key] has no records."
+	var/return_text = "<UL>"
```
List and join

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,274 @@
+		if(L.inactivity >= 6000)
+			continue
+		addtimer(L.update_exp_list(mins, ann),10)
+		CHECK_TICK
```
And this check tick won't be necessary once it's in the SS

Cyberboss commented on code/modules/mob/dead/new_player/login.dm:
```diff
@@ -1,4 +1,6 @@
 /mob/dead/new_player/Login()
+	if(config.use_exp_tracking)
+		client.set_exp_from_db()
```
Do this in client/New

Shadowlight213 replied:
That would cause the exp to be reset to the db value every time the client reconnected. The prefs datum survives a logout.

MrStonedOne replied:
>That would cause the exp to be reset to the db value every time the client reconnected.

That's the point.

We have two servers, you can't assume this data hasn't changed, it should be updated on the server when it changes, and it shouldn't be cached for too long.

Shadowlight213 replied:
Ok. with the new changes back to the old updating version, the exp is read from the db before each update. The only need to read without updating would be in the lobby for getting job unlock times

Shadowlight213 committed:
some stuff

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

Shadowlight213 committed:
Added button to toggle exempt status on a player.
Still need to make it update the db

Shadowlight213 committed:
REEE Jordie.
Undoes delaying exp updating to round end.
Adds the updating filtering to the update_exp_client proc which is used again

Shadowlight213 committed:
Lowers a bunch of the times needed.

Shadowlight213 committed:
Makes tracking per job instead of per department.
Department time is now calculated from job time.

Shadowlight213 commented:
OK. times are now tracked per job.
Departmental time is now calculated from the sum of the times for those jobs.

Unlock times for jobs have been heavily reduced
Captain: 3 hours crew
CMO: 3 hours as medical job
Geneticist: 1 hour crew
Chemist: 1 hours crew
Viro: 1 hours crew
HOS: 5 hours sec department
Sec officer: 5 hours crew
Warden: 5 hours crew
Detective: 5 hours crew
RD: 3 hours science department
Scientist: 1 hour crew
Roboticist: 1 hour crew
CE: 3 hours engineering
Engineer: 1 hour crew
Atmos Tech: 1 hour crew
HOP: 3 hours cargo
AI: 3 hours crew
Borg: 2 hours crew

Repukan commented:
When is the grace month kicking in?

Shadowlight213 commented:
Currently waiting on #27139 to be merged.

Shadowlight213 committed:
Whoops forgot to update master first. Lemme save this

Reverb should play cheeki breeki to people

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

Shadowlight213 committed:
I think I did this right.
Too tired to test atm.
Busy rewatching twintails ni narimasu

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+		return 0
+	if(!config.use_exp_restrictions_other && !(jobtitle in GLOB.command_positions))
+		return 0
+	return 1
```
FALSE/TRUE for returns

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+	if(!SSdbcore.Connect())
+		return -1
+	for(var/client/L in GLOB.clients)
+		if(L.inactivity >= 6000)
```
use the is_afk proc, i believe it has an inactivity param

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+	LAZYCLEARLIST(GLOB.exp_to_update)
+
+//Manual incrementing/updating
+/client/proc/update_exp_client(var/minutes, var/announce_changes = 0)
```
FALSE, also remove var/s

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+	else
+		return "0h"
+
+/proc/update_exp(var/mins, var/ann = 0)
```
Remove var/

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+	var/exp_living = text2num(prefs.exp[EXP_TYPE_LIVING])
+	return get_exp_format(exp_living)
+
+/proc/get_exp_format(var/expnum)
```
Remove var/

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+	usr << browse(body.Join(), "window=playerplaytime[C.ckey];size=550x615")
+
+
+/datum/admins/proc/toggle_exempt_status(var/client/C)
```
Remove var/

Cyberboss commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,313 @@
+		return
+
+
+/client/proc/update_exp_list(var/minutes, var/announce_changes = 0)
```
Remove var/ also FALSE

Cyberboss commented on code/modules/jobs/jobs.dm:
```diff
@@ -59,6 +59,24 @@ GLOBAL_LIST_INIT(nonhuman_positions, list(
 	"pAI"))
 
+GLOBAL_LIST_INIT(exp_jobsmap, list(
+	EXP_TYPE_CREW = list(titles = command_positions | engineering_positions | medical_positions | science_positions | supply_positions | security_positions | civilian_positions | list("AI","Cyborg")), // crew positions
```
This definition scares me for reasons

Shadowlight213 committed:
Removes var from proc arguments
Fixes database stuff

MrStonedOne commented on code/modules/jobs/job_exp.dm:
```diff
@@ -0,0 +1,325 @@
+		if(play_records[jtype] != old_records[jtype])
+			var jobname = jtype
+			var time = play_records[jtype]
+			var/datum/DBQuery/update_query = SSdbcore.NewQuery("INSERT INTO [format_table_name("role_time")] (`ckey`, `job`, `minutes`) VALUES ('[sanitizeSQL(ckey)]', '[sanitizeSQL(jobname)]', '[sanitizeSQL(time)]') ON DUPLICATE KEY UPDATE minutes = VALUES(minutes)")
```
I'd hazard a guess you forgot to remove this and use the new system?

Shadowlight213 replied:
It was supposed to be a manual proc call for debugging purposes. But since any db call gets auto denied from advanced proc call now, I should probably remove it.

MrStonedOne commented on SQL/tgstation_schema.sql:
```diff
@@ -254,6 +254,21 @@ CREATE TABLE `messages` (
+CREATE TABLE `role_time` 
+( `ckey` VARCHAR(32) NOT NULL ,
+ `job` VARCHAR(128) NOT NULL ,
+ `minutes` INT UNSIGNED NOT NULL,
```
So I did the math on this, (because it's my job).

In 8,378 years, we will have to upgrade this to a long.

Think we can manage that by then?

Jordie0608 replied:
We need to start early to avoid the Y10.395K bug.

kevinz000 commented:
>i was yelled at for even suggesting we do something like this

:(

Bawhoppen commented:
it's a bit too severe probably

Shadowlight213 commented:
@Bawhoppen Considering that a single round is about 30 minutes to an hour, playing 2 rounds will already unlock the base jobs. time wise.
Head of staff would be 3 to 6 rounds.
How the hell is this requiring 24/7 playtime?

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into playerexp

MrStonedOne commented:
Could you move exempt status to the database please, it could be a column of the player table, or more appropriate, give the player table a flags column and just do bitwise math.

Jordie0608 commented on SQL/database_changelog.txt:
```diff
@@ -1,3 +1,9 @@
+23rd April 2017, by Shadowlight213
+
+Added role_time table to track time spent playing departments. see the schema.
```
Just add the `CREATE TABLE` query here you nerd.

AlexCrimson commented:
IMO only head roles should be gated. Its frustrating to see shitty heads that have no idea what they are doing. Maybe Quartermaster too, although as its more of a pseudo-head role, it could have much lower time requirements. Captain should require X time spent as a head, or a combined time from multiple departments. 10 hours in one department to unlock a head role, then 20 hours total over at least two departments to unlock Captain. Something like that, although obviously the time requirements would need to be set by headmins.

Having other roles gated would hurt more than help. People who log on and see gated non-head roles will just leave.

If account stat tracking is a thing, players should be getting a tab to check their stats. Time played, times played as antag, kills, rounds survived etc.

KorPhaeron commented:
Did you manage to remove the proper player panel with this?

Shadowlight213 commented:
@KorPhaeron no, that bug has nothing to do with my pr, and happens even without it being testmerged.

Shadowlight213 committed:
Fixes conflicts

Shadowlight213 commented:
@MrStonedOne exempt status is tracked like a job, and stored alongside them in the database. Making a player flags column would require more db calls.

Shadowlight213 committed:
Does some requested changes.
Makes special_role be tracked instead of bundled under special.

Shadowlight213 committed:
Conflicts

MrStonedOne commented:
So, the issue with that layout shadow is that I don't have an easy way to do mass operations to remove and add exempt status to all players

Shadowlight213 commented:
Oh. I see. That makes sense.
I'll try to add it, but I've kinda run out of energy for this pr, so I dunno how long that might take.

DoctoreMad commented:
I support job locking.
I mean I'd hate if I was job locked but it would better than letting anyone play Head roles.

Then again, reading this PR comments makes me think people see Head roles in a different light

Bawhoppen commented:
Please get this done, like it is has always been needed, though I think it should be only restricting heads/silicions (maybe sec)

Tacolizard commented:
I support pretty heavy locking. I think that if people have to earn jobs then they will take them more seriously. I think that if you're jobbanned, you should have to re-earn the job you've been banned from.

Ausops commented:
/tg/ isn't the first server all new players come to when finding out about SS13 and while their own competence and empathy levels might vary, it's a bit silly to restrict basic jobs like doctor, scientist, engineer etc. Even the borg role, because playing for 2 hours as an assistant teaches you just about nothing on how to play a silicon role and by far the grief I've had to deal with involve humans or things with hands, not borgs.

AnturK commented:
Is this anywhere near technical completion, not even talking about how the locking should work.

Shadowlight213 commented:
Its really close actually. I just need to implement the flag thing in the player table mso wanted. Work on that is about 50% complete on my local repo.
Its been technically complete ,multiple times before, but then people keep asking for more stuff/requesting refactors.

I have my last final tomorrow, so I'll complete it afterwards.

lzimann commented:
@Shadowlight213

lzimann commented:
Ping if you want this reopen.

Shadowlight213 commented:
Ok. To clarify the state of the PR atm,
The system itself is done basically. I added @mrstonedone's flag column and it works and everything. Haven't pushed that yet since pr is closed.

The problem, that I discovered during testing, was that a lot of ghost roles, like ashwalkers, have their assigned role set to assistant. This is why you can see on ned's site here https://atlantaned.space/newSS13tools/stats/jobs.php
Assistant is way bigger than any of the other jobs, since all ghost roles were feeding into it.
Fixing that up is way harder than it seems, since it runs afoul of existing data integrity checks, so it'll take a while.
The hard part with a pr like this, is it has to be done right the first time, If junk data slips through, mso is the only one who can clear it, and it would throw off stuff like the percentage calculations for time playing that job vs total time alive.
It's kinda like the reverse of how goof says he can't be expected to test every single edge case before sending it out for a testmerge/merge, as well as if an admin touches it, the warranty is void.
For this, I kind of have to find and handle every edge case, and admins messing with stuff has to be dealt with correctly, before this goes on live, since the damage done is much harder to fix.

tldr version is it works, but data collection is flawed, making assistant data, and by extension time played as crew (for job unlocks) useless.

It'll take me some more time to figure out and refactor ghost role assigned roles.
